### DISCUSSION TITLE
Edge Functions: debug them locally with breakpoints

### DISCUSSION THREAD

**Discussion by christenbc** [supabase/supabase#13985] [General]
How can I debug properly with breakpoints a index.ts functions in VS Code?

**Comment by nyannyacha** [comment]
This feature is currently tracking on [PR-245](https://github.com/supabase/edge-runtime/pull/245) in Edge Runtime repository.

**Reply by christenbc** [reply]
After half a year, we got an answer!

Thanks for the reply.

**Comment by tecoad** [comment]
Anyone can supply an example launch.json for this?

**Reply by semics-tech** [reply]
@nyannyacha myself (and I'm sure lots of others) would be very greatful if you could provide an example of how to set this up with vscode?

**Reply by nyannyacha** [reply]
Hello @semics-tech 😁

This is basically using v8 inspector protocol (DevTools), so you should be able to connect to the worker inspector session in the edge runtime via [VsCode Node Debugger Adapter](https://code.visualstudio.com/docs/nodejs/nodejs-debugging).

Below is a VsCode launch template that you can attach the debugger to a worker of the edge runtime.

```jsonc
{
    "name": "Attach V8 Inspector",
    "type": "node",
    "address": "127.0.0.1",
    "request": "attach",

    // This is important: You must enter the path where your edge functions
    // folders are located.
    "localRoot": "${workspaceFolder}/supabase/functions/",

    // CLI mounts all functions under /home/deno/functions
    "remoteRoot": "/home/deno/functions/",

    // As you can see from the two options above, VsCode tries to locate the script
    // paths being used by the inspector session locally by matching local and remote
    // paths.
    //
    // If a valid path cannot be found on either side, or the script contents of
    // those paths do not match each other, the breakpoint setting may not proceed
    // correctly.
    //
    // We set the two options below for this.
    "sourceMaps": true,
    "enableContentValidation": false,

    // Properties below are optional.
    "restart": true,
    "timeout": 1000000,

    // Inspector is accepting sessions on 8083 by default.
    "port": 8083
}
```

Also, before attaching to an inspector session in VsCode, you should check through the [Supabase CLI](https://github.com/supabase/cli) that the locally running edge runtime is in a state to accept an inspector session, and if not, [pass a flag to the CLI](https://github.com/supabase/cli/blob/develop/docs/supabase/functions/serve.md) to enable the inspector capability.

You may want to check out the link below for instructions on how to enable the edge runtime inspector capability.

https://supabase.com/docs/guides/functions/debugging-tools

**Reply by semics-tech** [reply]
@nyannyacha thankyou so much that worked (with a slight change)! I had to remove the `"remoteRoot": "/home/deno/functions/",`  key and it started working.

I do have a slightly annoying issue where it keeps hitting a breakpoint that I didn't set in a package, have you seen this before? if not, I'll do some research as it's likely an issue with my setup.

![image](https://github.com/user-attachments/assets/f543b767-c18b-496b-b285-e28a6556f4b7)


Also to anyone else who's coming to this, you need run it in this order
1. `supabase functions serve --inspect-mode brk`.
2. Start the `Attach V8 Inspector` debug process.
3. Add a breakpoint to your code.

**Reply by MattBrown88** [reply]
The above wasn't working for me. Here's the launch.json that did work for me. 

```
 {
      "name": "Debug Edge Functions",
      "type": "node",
      "request": "attach",
      "address": "127.0.0.1",
      "port": 8083,
      "sourceMaps": true,
      "enableContentValidation": false,
      "restart": true,
      "timeout": 1000000,
      "sourceMapPathOverrides": {
        "file:///var/tmp/sb-compile-edge-runtime/*": "${workspaceFolder}/supabase/functions/*"
      }
    }
```

The key parameter to be aware of is "sourceMapPathOverrides". You may need to update the left-hand side to match the compiled edge runtime path in your environment:

`file:///var/tmp/sb-compile-edge-runtime/*`

I wrote up the full set of steps here in case it helps anyone else:

https://medium.com/@matt_84360/debug-supabase-edge-functions-in-vs-code-with-breakpoints-fabe8582fcf1
